### ISSUE TITLE
Does not run on ARM/raspberry pi: mdb_env_open: cannot allocate memory

### ISSUE THREAD

**Issue by girino** [bitvora/wot-relay#9]
I am trying to run this on a raspberry pi, but as soon as i run it, i get this memory allocation problema. (this seems to be a problem of lmdb on arm processors. I hat trouble with strfry running on the rpi too, and had to reduce the "mapsize" param in the confs.

I checked code for github.com/fiatjaf/eventstore/lmdb and it seems that you have "env.SetMapSize(1 << 38) // ~273GB". This might be causing the problem in the rpi.

Error message:

wot-relay  | starting
wot-relay  | panic: mdb_env_open: cannot allocate memory
wot-relay  |
wot-relay  | goroutine 1 [running]:
wot-relay  | main.main()
wot-relay  |    /app/main.go:55 +0x7e8
wot-relay  |
wot-relay exited with code 2

**Comment by barrydeen** [comment]
cc @fiatjaf

**Comment by fiatjaf** [comment]
LMDB doesn't run everywhere. You can release two builds, one with LMDB and the other with Badger.

**Comment by girino** [comment]
I forked github.com/fiatjaf/eventstore/lmdb to allow for MapSize as a parameter to LMDBBackend, and could make it work with lower values of MapSize (tested with 100000000000 and it's working right now.

```
diff --git a/main.go b/main.go
index ba3ce7f..b996c52 100644
--- a/main.go
+++ b/main.go
@@ -11,7 +11,7 @@ import (
        "time"

        "github.com/cespare/xxhash"
-       "github.com/fiatjaf/eventstore/lmdb"
+       "github.com/girino/eventstore/lmdb"
        "github.com/fiatjaf/khatru"
        "github.com/greatroar/blobloom"
        "github.com/joho/godotenv"
@@ -50,6 +50,8 @@ func main() {

        db := lmdb.LMDBBackend{
                Path: getEnv("DB_PATH"),
+                MaxLimit: 0,
+                MapSize: 100000000000,
        }
        if err := db.Init(); err != nil {
                panic(err)
```

and in github.com/fiatjaf/eventstore:

```
diff --git a/lmdb/lib.go b/lmdb/lib.go
index 291a275..6d303e4 100644
--- a/lmdb/lib.go
+++ b/lmdb/lib.go
@@ -19,6 +19,7 @@ var _ eventstore.Store = (*LMDBBackend)(nil)
 type LMDBBackend struct {
        Path     string
        MaxLimit int
+       MapSize int64

        lmdbEnv *lmdb.Env

@@ -49,7 +50,11 @@ func (b *LMDBBackend) Init() error {

        env.SetMaxDBs(10)
        env.SetMaxReaders(1000)
-       env.SetMapSize(1 << 38) // ~273GB
+       if b.MapSize == 0 {
+               env.SetMapSize(1 << 38) // ~273GB
+       } else {
+               env.SetMapSize(b.MapSize)
+       }

        // create directory if it doesn't exist and open it
        if err := os.MkdirAll(b.Path, 0755); err != nil {
```

@fiatjaf if you could accept MapSize as part of LMDBBackend  or as a parameter somehow, this could be easily fixed in wot-relay using an env. variable or config.

**Comment by girino** [comment]
I tried with badger. The change was quite simple, and works like a charm on the raspberry pi:

diff --git a/main.go b/main.go
index 589f848..73afad2 100644
--- a/main.go
+++ b/main.go
@@ -11,7 +11,7 @@ import (
        "time"

        "github.com/cespare/xxhash"
-       "github.com/fiatjaf/eventstore/lmdb"
+       "github.com/fiatjaf/eventstore/badger"
        "github.com/fiatjaf/khatru"
        "github.com/greatroar/blobloom"
        "github.com/joho/godotenv"
@@ -48,7 +48,7 @@ func main() {
        relay.Info.Description = config.RelayDescription
        appendPubkey(config.RelayPubkey)

-       db := lmdb.LMDBBackend{
+       db := badger.BadgerBackend{
                Path: getEnv("DB_PATH"),
        }
        if err := db.Init(); err != nil {

**Comment by barrydeen** [comment]
would be awesome if you could submit a PR where the user can pick the db type in the .env

**Comment by girino** [comment]
> would be awesome if you could submit a PR where the user can pick the db type in the .env

I'm actually working on a compile time selection of the DB, since the dbs are platform dependent. Will post a PR as soon as i tested it.

**Comment by barrydeen** [comment]
Fixed in #13

**Comment by girino** [comment]
Reopening this issue, since the use of badger is not a solution. It is VERY memory consuming (2GB after a few hours, had to disable swap to reduce cpu wait time, lmdb never goes over 300 Mb). Submitted a pull request to fiatjaf/eventstore, if it gets accepted i'll add a pull request here adding an environment variable to allow for lower MapSizes in lmdb
